File: src/main/java/au/id/micolous/metrodroid/key/ClassicStaticKeys.java
/*
 * ClassicStaticKeys.java
 *
 * Copyright 2018 Google
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.key;

import android.annotation.SuppressLint;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import au.id.micolous.metrodroid.util.Utils;

/**
 * Helper for access to static MIFARE Classic keys. This can be used for keys that should be
 * attempted on multiple cards.
 *
 * This is only really useful when a transit agency doesn't implement key diversification.
 *
 * See https://github.com/micolous/metrodroid/wiki/Importing-MIFARE-Classic-keys#static-json for
 * a file format description.
 */
public class ClassicStaticKeys extends ClassicCardKeys {
    private static final String SECTOR_IDX = "sector";
    public static final String JSON_TAG_ID_DESC = "Description";

    // Sparse Arrays are annoying to iterate through
    @SuppressLint("UseSparseArrays")
    private Map<Integer, List<ClassicSectorKeyWrapper>> mKeys = new HashMap<>();
    private String mDescription;

    public void mergeJSON(JSONObject string) throws JSONException {
        merge(fromJSON(string));
    }

    private void merge(ClassicStaticKeys classicStaticKeys) {
        for (Map.Entry<Integer, List<ClassicSectorKeyWrapper>> cl : classicStaticKeys.mKeys.entrySet()) {
            if (!mKeys.containsKey(cl.getKey()))
                mKeys.put(cl.getKey(), new ArrayList<>());
            mKeys.get(cl.getKey()).addAll(cl.getValue());
        }
    }

    public String getDescription() {
        return mDescription;
    }

    private static class ClassicSectorKeyWrapper extends ClassicSectorKey {
        private int mSectorIndex;

        private ClassicSectorKeyWrapper() {}

        public static ClassicSectorKeyWrapper fromJSON(JSONObject json) throws JSONException {
            ClassicSectorKeyWrapper w = new ClassicSectorKeyWrapper();
            w.mSectorIndex = json.getInt(SECTOR_IDX);
            fromJSON(w, json);
            return w;
        }

        public JSONObject toJSON() throws JSONException {
            JSONObject o = super.toJSON();
            o.put(SECTOR_IDX, mSectorIndex);
            return o;
        }

        int getSectorIndex() {
            return mSectorIndex;
        }
    }

    public JSONObject toJSON() throws JSONException {
        JSONArray keysJson = new JSONArray();
        for (Map.Entry<Integer, List<ClassicSectorKeyWrapper>> sector : mKeys.entrySet()) {
            for (ClassicSectorKey key : sector.getValue()) {
                keysJson.put(key.toJSON());
            }
        }

        JSONObject json = new JSONObject();
        json.put(KEYS, keysJson);
        if (mDescription != null)
            json.put(JSON_TAG_ID_DESC, mDescription);
        return json;
    }

    public static ClassicStaticKeys fromJSON(JSONObject jsonRoot) throws JSONException {
        ClassicStaticKeys ins = new ClassicStaticKeys();
        ins.mDescription = jsonRoot.optString(JSON_TAG_ID_DESC);
        JSONArray keysJSON = jsonRoot.getJSONArray(KEYS);
        for (int i = 0; i < keysJSON.length(); i++) {
            JSONObject json = keysJSON.getJSONObject(i);

            ClassicSectorKeyWrapper w = ClassicSectorKeyWrapper.fromJSON(json);

            if (!ins.mKeys.containsKey(w.getSectorIndex())) {
                ins.mKeys.put(w.getSectorIndex(), new ArrayList<>());
            }

            ins.mKeys.get(w.getSectorIndex()).add(w);
        }

        ins.setLengthAndReturn(jsonRoot);
        return ins;
    }

    @Override
    @NonNull
    public List<ClassicSectorKey> getCandidates(int sectorIndex) {
        if (!mKeys.containsKey(sectorIndex))
            return Collections.emptyList();
        return (List<ClassicSectorKey>)(List<?>)mKeys.get(sectorIndex);
    }

    @NonNull
    @Override
    public List<ClassicSectorKey> keys() {
        List<ClassicSectorKey> allKeys = new ArrayList<>();
        for (List<ClassicSectorKeyWrapper> keys : mKeys.values()) {
            allKeys.addAll(keys);
        }
        return allKeys;
    }

    @Override
    public String getType() {
        return CardKeys.TYPE_MFC_STATIC;
    }

    @Nullable
    @Override
    public String getUID() {
        return CLASSIC_STATIC_TAG_ID;
    }
}

File: src/main/java/au/id/micolous/metrodroid/key/ClassicStaticKeys.kt
/*
 * ClassicStaticKeys.java
 *
 * Copyright 2018 Google
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.key

import android.content.Context
import android.database.Cursor
import au.id.micolous.farebot.R
import au.id.micolous.metrodroid.provider.KeysTableColumns
import au.id.micolous.metrodroid.util.Utils
import org.json.JSONException
import org.json.JSONObject

/**
 * Helper for access to static MIFARE Classic keys. This can be used for keys that should be
 * attempted on multiple cards.
 *
 * This is only really useful when a transit agency doesn't implement key diversification.
 *
 * See https://github.com/micolous/metrodroid/wiki/Importing-MIFARE-Classic-keys#static-json for
 * a file format description.
 */
class ClassicStaticKeys private constructor(override val description: String?,
                                            keys: Map<Int, List<ClassicSectorKey>>,
                                            override val sourceDataLength: Int)
    : ClassicKeysImpl(mKeys = keys) {

    private operator fun plus(other: ClassicStaticKeys): ClassicStaticKeys {
        return ClassicStaticKeys(description = description,
                keys = ClassicKeysImpl.flattenKeys(listOf(this, other)),
                sourceDataLength = sourceDataLength + other.sourceDataLength)
    }

    @Throws(JSONException::class)
    override fun toJSON(): JSONObject {
        val json = baseJson
        if (description != null)
            json.put(JSON_TAG_ID_DESC, description)
        return json
    }

    override val type = CardKeys.TYPE_MFC_STATIC

    override val uid = CardKeys.CLASSIC_STATIC_TAG_ID

    override val fileType: String
        get() = Utils.localizePlural(R.plurals.keytype_mfc_static, keyCount, keyCount)

    companion object {
        private const val JSON_TAG_ID_DESC = "Description"

        /**
         * Retrieves all statically defined MIFARE Classic keys.
         * @return All [ClassicCardKeys], or null if not found
         */
        fun forStaticClassic(context: Context, retriever: CardKeysRetriever): ClassicStaticKeys? {
            val cur = retriever.makeStaticClassicCursor(context) ?: return fallback()
            return fromCursor(cur)
        }

        private fun fromCursor(cursor: Cursor): ClassicStaticKeys? {
            var keys: ClassicStaticKeys? = null

            // Static key requests should give all of the static keys.
            while (cursor.moveToNext()) {
                if (cursor.getString(cursor.getColumnIndex(KeysTableColumns.CARD_TYPE)) != CardKeys.TYPE_MFC_STATIC)
                    continue
                try {
                    val id = cursor.getInt(cursor.getColumnIndex(KeysTableColumns._ID))
                    val json = JSONObject(cursor.getString(cursor.getColumnIndex(KeysTableColumns.KEY_DATA)))
                    val nk = ClassicStaticKeys.fromJSON(json, "cursor/$id") ?: continue
                    if (keys == null)
                        keys = nk
                    else
                        keys += nk
                } catch (ignored: JSONException) {
                }
            }
            return keys
        }

        fun fallback() = ClassicStaticKeys(description = "fallback",
                keys = mapOf(), sourceDataLength = 0)

        @Throws(JSONException::class)
        fun fromJSON(jsonRoot: JSONObject, defaultBundle: String) = try {
            ClassicStaticKeys(
                    description = jsonRoot.optString(JSON_TAG_ID_DESC),
                    keys = ClassicKeysImpl.keysFromJSON(jsonRoot, false, defaultBundle),
                    sourceDataLength = jsonRoot.toString().length)
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }
}
